tree-walk.c

```diff
@@ -521,19 +529,28 @@ static int find_tree_entry(struct tree_desc *t, const char *name, struct object_
 			oidcpy(result, oid);
 			return 0;
 		}
-		return get_tree_entry(oid, name + entrylen, result, mode);
+		return get_tree_entry_gently(oid, name + entrylen, result, mode, gently);
 	}
 	return -1;
 }
 
-int get_tree_entry(const struct object_id *tree_oid, const char *name, struct object_id *oid, unsigned *mode)
+int get_tree_entry_gently(const struct object_id *tree_oid, const char *name,
+			  struct object_id *oid, unsigned *mode, int gently)
 {
 	int retval;
 	void *tree;
 	unsigned long size;
 	struct object_id root;
 
-	tree = read_object_with_reference(tree_oid, tree_type, &size, &root);
+	if (gently) {
+		struct object_info oi = OBJECT_INFO_INIT;
+
+		if (oid_object_info_extended(the_repository, tree_oid, &oi, 0) < 0)
+			return -1;
+		tree = *oi.contentp;
+	} else {
+		tree = read_object_with_reference(tree_oid, tree_type, &size, &root);
+	}
```

Comment:
A good rule to follow is to handle the easier code block first. This not only resembles natural human behavior 😄 but it also makes it easier to follow the logic for the occasional reader:

```c
if (!gently)
	tree = read_object_with_reference(tree_oid, tree_type, &size, &root);
else {
	struct object_info oi = OBJECT_INFO_INIT;

	if (oid_object_info_extended(the_repository, tree_oid, &oi, 0) < 0)
		return -1;
	tree = *oi.contentp;
}
```